src/librustdoc/test.rs

```diff
@@ -230,11 +233,28 @@ fn run_test(
     }
 
     let outdir = if let Some(mut path) = options.persist_doctests {
+        // For example `module/file.rs` would become `module_file_rs`
+        let folder_name = filename
+            .to_string()
+            .chars()
+            .map(|c| if c == '/' || c == '.' { '_' } else { c })
+            .collect::<String>();
+
         path.push(format!(
-            "{}_{}",
-            filename.to_string().rsplit('/').next().unwrap().replace(".", "_"),
-            line
+            "{name}_{number}",
+            name = folder_name,
+            number = {
+                // Increases the current test number, if this file already
+                // exists or it creates a new entry with a test number of 0.
+                visited_tests
+                    .lock()
+                    .unwrap()
+                    .entry(folder_name.clone())
+                    .and_modify(|v| *v += 1)
+                    .or_insert(0)
```

Comment:
This part would then be changed to something like this:

```rust
let mut hasher = std::collections::hash_map::DefaultHasher::new();
folder_name.hash(&mut hasher);

visited_tests
    .lock()
    .unwrap()
    .entry(hasher.finish())
    .and_modify(|v| *v += 1)
    .or_insert(0)
```

which would most likely increase performance, because only 8 bytes have to be saved for each test in the `HashMap`, instead of an arbitrary number of bytes, which would be most likely larger than 8 bytes. I think cloning an entire `String` should be slower than hashing it.

I am not sure if this change is desirable, because it might make the code harder to read?